html/webappapis/dynamic-markup-insertion/opening-the-input-stream/tasks.window.js

```diff
@@ -0,0 +1,114 @@
+// is no longer executed.
+//
+// This file comprehensively (but not exhaustively) tests for many queued tasks
+// that may be observable. Each taskTest() call in fact runs two tests: the
+// first one "tasks without document.open()" does not actually run
+// document.open(), just to test that the tested task works ordinarily; the
+// second actually calls document.open() to test if the method call removes
+// that specific task from the queue.
+
+setup({
+  allow_uncaught_exception: true
+});
+
+function taskTest(description, testBody) {
+  async_test(t => {
+    const frame = document.body.appendChild(document.createElement("iframe"));
+    // The empty HTML seems to be necessary to cajole Chrome into firing a load
+    // event, which is necessary to make sure the frame's document doesn't have
+    // a parser associated with it.
+    frame.src = "/common/blank.html";
+    t.add_cleanup(() => frame.remove());
+    frame.onload = t.step_func(() => {
+      // Make sure there is no parser. Firefox seems to have an additional
+      // non-spec-compliant readiness state "uninitialized", so test for the
+      // two known valid readiness states instead.
+      assert_in_array(frame.contentDocument.readyState, ["interactive", "complete"]);
+      testBody(t, frame, doc => {});
+    });
+  }, `tasks without document.open() (${description})`);
+
+  async_test(t => {
+    const frame = document.body.appendChild(document.createElement("iframe"));
+    // The empty HTML seems to be necessary to cajole Chrome into firing a load
+    // event, which is necessary to make sure the frame's document doesn't have
+    // a parser associated with it.
+    frame.src = "/common/blank.html";
+    t.add_cleanup(() => frame.remove());
+    frame.onload = t.step_func(() => {
+      // Make sure there is no parser. Firefox seems to have an additional
+      // non-spec-compliant readiness state "uninitialized", so test for the
+      // two known valid readiness states instead.
+      assert_in_array(frame.contentDocument.readyState, ["interactive", "complete"]);
+      testBody(t, frame, doc => doc.open());
+    });
+  }, `document.open() and tasks (${description})`);
+}
+
+taskTest("timeout", (t, frame, open) => {
+  let happened = false;
+  // Work around the lint script, since we can't use step_timeout here.
+  frame.contentWindow['setTimeout'](() => happened = true, 100);
+  open(frame.contentDocument);
+  t.step_timeout(() => {
+    assert_true(happened);
+    t.done();
+  }, 200);
+});
+
+taskTest("window message", (t, frame, open) => {
+  let counter = 0;
+  frame.contentWindow.postMessage(undefined, "*");
+  open(frame.contentDocument);
+  frame.contentWindow.postMessage(undefined, "*");
+  frame.contentWindow.onmessage = t.step_func(e => {
+    assert_equals(e.data, undefined);
+    counter++;
+    assert_less_than_equal(counter, 2);
+    if (counter == 2) {
+      t.done();
+    }
+  });
+});
+
+taskTest("canvas.toBlob()", (t, frame, open) => {
```

Comment:
The "queue a task" of `toBlob` happens inside "in parallel" so might not have happened when `document.open()` is called, is that a problem for this test? Does any browser currently fail this?

Reply:
Good point. No browser actually seems to fail this, but I should note that `setTimeout` is defined the same way (wait in parallel, and then queue a task), and browsers have very different behavior with regards to that. That's why I think this is a good test.